py3status/modules/speedtest_cli.py

```diff
@@ -0,0 +1,294 @@
+    format += ' [\?color=download ↓ {download} {download_unit}]'
+}
+
+# colored based on comparation between two last runs
+speedtest_cli {
+    format = ' [[\?if=download_raw<previous_download_raw&color=degraded ↓]|[\?color=ok ↑]]'
+}
+
+# compare only download
+speedtest_cli {
+    format = '[[\?if=quality=slower&color=download ↓ {download} {download_unit}]'
+    format += '[\?if=quality=faster&color=download ↑ {download} {download_unit}]'
+    format += '[\?if=quality=ok&color=download -> {download} {download_unit}]'
+    format += '[\?if=quality=bad&color=download x {download} {download_unit}]]'
+    format += '|'
+}
+```
+
+@author Cyril Levis (@cyrinux)
+
+SAMPLE OUTPUT
+{'full_text': u'\u25cf 208.18 ms \u2191 0.247 MB/s \u2193 0.453 MB/s'}
+"""
+
+from json import loads
+
+STRING_NOT_INSTALLED = "not installed"
+MISSING_PLACEHOLDER = "{download} or {upload} placeholder required"
+
+
+class Py3status:
+    """
+    """
+
+    # available configuration parameters
+    button_refresh = 1
+    button_share = None
+    format = (
+        u"\u25cf [{ping} ms] [↑ {upload} {upload_unit}] [↓ {download} {download_unit}]"
+    )
+    server_id = None
+    si_units = False
+    sleep_timeout = 5
+    thresholds = {
+        "download": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+        "ping": [(200, "bad"), (150, "orange"), (100, "degraded"), (10, "good")],
+        "quality": [
+            ("ok", "good"),
+            ("bad", "degraded"),
+            ("faster", "good"),
+            ("slower", "degraded"),
+            ("faster", "good"),
+        ],
+        "upload": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+    }
+    timeout = 10
+    unit_bitrate = "MB/s"
+    unit_size = "MB"
+
+    def post_config_hook(self):
+        if not self.py3.check_commands("speedtest-cli"):
+            raise Exception(STRING_NOT_INSTALLED)
+
+        self.first_run = True
+        self.placeholders = list(set(self.py3.get_placeholders_list(self.format)))
+
+        dont_upload = ''
+        if 'upload' not in self.placeholders:
+            dont_upload = '--no-upload'
+
+        dont_download = ''
+        if 'download' not in self.placeholders:
+            dont_download = '--no-download'
+
+        server = ''  # if not specified, nearest server is use
+        if self.server_id and int(self.server_id):
+            server = '--server {}'.format(self.server_id)
+
+        share = ''
+        if self.button_share and all(
+            x in self.placeholders for x in ['download', 'upload']
+        ):
+            share = '--share'
+
+        self.command = "speedtest-cli --json --secure --timeout {} {} {} {} {}".format(
+            self.timeout, dont_upload, dont_download, server, share
+        )
+
+        # fail if download or upload missing
+        if not any(x in ['download', 'upload'] for x in self.placeholders):
+            raise Exception('%s' % (MISSING_PLACEHOLDER))
```

Comment:
Throw this exception first thing first so you don't have to do some unnecessary work.

Reply:
True

Reply:
I changed the behavior